File: app/android/app/src/main/java/com/ambientnotes/app/recognition/providers/AcrCloudProvider.kt
package com.ambientnotes.app.recognition.providers

import com.ambientnotes.app.recognition.RecognitionProvider
import com.ambientnotes.app.recognition.RecognitionProviderException
import com.ambientnotes.app.recognition.RecognitionResult
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.withContext
import okhttp3.MediaType.Companion.toMediaType
import okhttp3.MultipartBody
import okhttp3.OkHttpClient
import okhttp3.Request
import okhttp3.RequestBody.Companion.toRequestBody
import org.json.JSONObject
import java.util.Base64
import javax.crypto.Mac
import javax.crypto.spec.SecretKeySpec

/**
 * Talks directly to the user's own ACRCloud project (https://www.acrcloud.com).
 * The user supplies their host/access key/access secret in Settings -- this
 * app never bundles or shares ACRCloud credentials; it's strictly BYO-account,
 * per the project requirements.
 */
class AcrCloudProvider(
    private val credentialsProvider: suspend () -> AcrCloudCredentials?,
    private val httpClient: OkHttpClient = OkHttpClient(),
) : RecognitionProvider {

    override val id: String = "acrcloud"
    override val displayName: String = "ACRCloud"

    override suspend fun isConfigured(): Boolean = credentialsProvider() != null

    override suspend fun recognize(pcm16Audio: ByteArray, sampleRateHz: Int): RecognitionResult =
        withContext(Dispatchers.IO) {
            val creds = credentialsProvider()
                ?: return@withContext RecognitionResult.noMatch(id)

            val timestamp = (System.currentTimeMillis() / 1000).toString()
            val signature = buildSignature(creds.accessSecret, creds.accessKey, timestamp)

            val body = MultipartBody.Builder()
                .setType(MultipartBody.FORM)
                .addFormDataPart("access_key", creds.accessKey)
                .addFormDataPart("sample_bytes", pcm16Audio.size.toString())
                .addFormDataPart("timestamp", timestamp)
                .addFormDataPart("signature", signature)
                .addFormDataPart("data_type", "audio")
                .addFormDataPart("signature_version", "1")
                .addFormDataPart(
                    "sample", "sample.wav",
                    pcm16Audio.toRequestBody("audio/wav".toMediaType())
                )
                .build()

            val request = Request.Builder()
                .url(buildIdentifyUrl(creds.host))
                .post(body)
                .build()

            val response = try {
                httpClient.newCall(request).execute()
            } catch (e: java.io.IOException) {
                throw RecognitionProviderException(id, "Network error contacting ACRCloud", e)
            }

            response.use {
                if (!it.isSuccessful) {
                    throw RecognitionProviderException(id, "ACRCloud HTTP ${it.code}")
                }
                val json = JSONObject(it.body?.string().orEmpty())
                parseResponse(json)
            }
        }

    /** Builds the identify endpoint URL. `host` is normally a bare hostname
     * (e.g. "identify-eu-west-1.acrcloud.com"), in which case we assume TLS.
     * A host already containing a scheme (used by tests, pointed at a local
     * MockWebServer over plain HTTP) is used as-is. */
    internal fun buildIdentifyUrl(host: String): String =
        if (host.startsWith("http://") || host.startsWith("https://")) {
            "${host.trimEnd('/')}/v1/identify"
        } else {
            "https://$host/v1/identify"
        }

    internal fun parseResponse(json: JSONObject): RecognitionResult {
        val statusCode = json.optJSONObject("status")?.optInt("code", -1) ?: -1
        if (statusCode != 0) return RecognitionResult.noMatch(id)

        val music = json.optJSONObject("metadata")
            ?.optJSONArray("music")
            ?.optJSONObject(0) ?: return RecognitionResult.noMatch(id)

        val artists = music.optJSONArray("artists")
        val artistNames = buildList {
            if (artists != null) {
                for (i in 0 until artists.length()) add(artists.optJSONObject(i)?.optString("name").orEmpty())
            }
        }.filter { it.isNotBlank() }.joinToString(", ")

        val externalIds = mutableMapOf<String, String>()
        music.optJSONObject("external_ids")?.let { ids ->
            ids.keys().forEach { key -> externalIds[key] = ids.optString(key) }
        }

        return RecognitionResult(
            matched = true,
            title = music.optString("title").takeIf { it.isNotBlank() },
            artist = artistNames.takeIf { it.isNotBlank() },
            album = music.optJSONObject("album")?.optString("name"),
            releaseDate = music.optString("release_date").takeIf { it.isNotBlank() },
            confidence = (music.optDouble("score", 0.0) / 100.0).toFloat(),
            providerName = id,
            externalIds = externalIds,
        )
    }

    private fun buildSignature(secret: String, accessKey: String, timestamp: String): String {
        val stringToSign = listOf("POST", "/v1/identify", accessKey, "audio", "1", timestamp)
            .joinToString("\n")
        val mac = Mac.getInstance("HmacSHA1")
        mac.init(SecretKeySpec(secret.toByteArray(Charsets.UTF_8), "HmacSHA1"))
        return Base64.getEncoder().encodeToString(mac.doFinal(stringToSign.toByteArray(Charsets.UTF_8)))
    }
}

data class AcrCloudCredentials(
    val host: String,
    val accessKey: String,
    val accessSecret: String,
)

File: service/app/providers/acrcloud_provider.py
"""
Optional cloud fallback using the *server operator's* ACRCloud project
(distinct from the per-user ACRCloud account the Android app can connect to
directly). Useful if the self-hoster wants "local library first, ACRCloud
for anything else" without asking every user to bring their own key.
"""
from __future__ import annotations

import base64
import hashlib
import hmac
import logging
import time

import httpx

from app.config import Settings
from app.models import RecognitionResult
from app.providers.base import RecognitionProvider
from app.providers.registry import register_provider

logger = logging.getLogger(__name__)


@register_provider("acrcloud")
class AcrCloudProvider(RecognitionProvider):
    def __init__(self, host: str, access_key: str, access_secret: str, enabled: bool = True):
        super().__init__()
        self.host = host
        self.access_key = access_key
        self.access_secret = access_secret
        self.enabled = enabled

    @classmethod
    def from_settings(cls, settings: Settings) -> "AcrCloudProvider":
        return cls(
            host=settings.acrcloud_host,
            access_key=settings.acrcloud_access_key,
            access_secret=settings.acrcloud_access_secret,
            enabled=settings.acrcloud_enabled,
        )

    async def is_available(self) -> bool:
        return bool(self.enabled and self.host and self.access_key and self.access_secret)

    def _build_signature(self, timestamp: str) -> str:
        string_to_sign = "\n".join(
            ["POST", "/v1/identify", self.access_key, "audio", "1", timestamp]
        )
        sig = hmac.new(
            self.access_secret.encode("utf-8"),
            string_to_sign.encode("utf-8"),
            digestmod=hashlib.sha1,
        ).digest()
        return base64.b64encode(sig).decode("utf-8")

    async def recognize(self, audio_bytes: bytes, sample_rate_hz: int) -> RecognitionResult:
        if not await self.is_available():
            return RecognitionResult(matched=False, provider=self.name)

        timestamp = str(int(time.time()))
        signature = self._build_signature(timestamp)
        data = {
            "access_key": self.access_key,
            "sample_bytes": str(len(audio_bytes)),
            "timestamp": timestamp,
            "signature": signature,
            "data_type": "audio",
            "signature_version": "1",
        }
        files = {"sample": ("sample.wav", audio_bytes, "audio/wav")}

        async with httpx.AsyncClient(timeout=15) as client:
            resp = await client.post(f"https://{self.host}/v1/identify", data=data, files=files)
            resp.raise_for_status()
            payload = resp.json()

        status = payload.get("status", {})
        if status.get("code") != 0:
            return RecognitionResult(matched=False, provider=self.name, raw_provider_response=payload)

        music = payload.get("metadata", {}).get("music", [{}])[0]
        return RecognitionResult(
            matched=True,
            title=music.get("title"),
            artist=", ".join(a.get("name", "") for a in music.get("artists", [])),
            album=music.get("album", {}).get("name"),
            release_date=music.get("release_date"),
            confidence=float(music.get("score", 0)) / 100.0,
            provider=self.name,
            external_ids={k: str(v) for k, v in music.get("external_ids", {}).items()},
            raw_provider_response=payload,
        )
